### PR TITLE
Fail to explicitly tail call on x86 unix.

### LINKED ISSUE #25020: [RyuJIT/Linux/x86] Tail-call optimization leads to assert on Linux/x86
Following example crashes due to assert in `morph.cpp` on Linux/x86:
```cs
using System;
using System.Reflection;
using System.Reflection.Emit;

namespace jitx86
{
    class Program
    {    
        static void Main(string[] args)
        {
            DynamicMethod dm = new DynamicMethod("MyMethod", typeof(string), new Type[] { typeof(string), typeof(string) });
            
            ILGenerator generator = dm.GetILGenerator();
            generator.Emit(OpCodes.Ldarg_0);
            generator.Emit(OpCodes.Ldarg_1);
            generator.Emit(OpCodes.Tailcall);
            generator.EmitCall(OpCodes.Call, typeof(String).GetMethod("Concat", new Type[] { typeof(string), typeof(string) }), null);
            generator.Emit(OpCodes.Ret);

            string a = "1234";
            string b = "abcd";
            
            Console.WriteLine(dm.Invoke(null, BindingFlags.Default, null, new object[] {a, b}, null));
        }
    }
}
```
```
$ ./corerun jitx86.dll
...
Morphing BB01 of 'DynamicClass:MyMethod(ref,ref):ref'

fgMorphTree BB01, stmt 1 (before)
               [000006] --C-G-------              *  RETURN    ref   
               [000003] --C-G-------              \--*  CALL      ref    System.String.Concat
               [000001] ------------ arg0            +--*  LCL_VAR   ref    V00 arg0         
               [000002] ------------ arg1            \--*  LCL_VAR   ref    V01 arg1         
fgMorphTailCall (before):
               [000003] --CXG-------              *  CALL      ref    System.String.Concat
               [000001] ------------ arg0         +--*  LCL_VAR   ref    V00 arg0         
               [000002] ------------ arg1         \--*  LCL_VAR   ref    V01 arg1         
fgMorphTailCall (after):
               [000003] --CXG-------              *  CALL      ref    System.String.Concat
               [000001] ------------ arg0         +--*  LCL_VAR   ref    V00 arg0         
               [000002] ------------ arg1         +--*  LCL_VAR   ref    V01 arg1         
               [000008] ------------ arg2         +--*  CNS_INT   int    0
               [000010] ------------ arg3         +--*  CNS_INT   int    9
               [000012] ------------ arg4         +--*  CNS_INT   int    8
               [000014] ------------ arg5         \--*  CNS_INT   int    7

GTF_CALL_M_TAILCALL bit set for call [000003]
New Basic Block BB02 [0002] created.
New Basic Block BB03 [0003] created.

Removing statement [000007] in BB01 as useless:
               [000007] ------------              *  STMT      void  (IL 0x000...0x004)
               [000003] --CXG-------              \--*  CALL      void   System.String.Concat
               [000001] ------------ arg0            +--*  LCL_VAR   ref    V00 arg0         
               [000002] ------------ arg1            +--*  LCL_VAR   ref    V01 arg1         
               [000008] ------------ arg2            +--*  CNS_INT   int    0
               [000010] ------------ arg3            +--*  CNS_INT   int    9
               [000012] ------------ arg4            +--*  CNS_INT   int    8
               [000014] ------------ arg5            \--*  CNS_INT   int    7

BB01 becomes empty
*** creating inlined GC Poll in top block BB01
               [000023] ------------              *  STMT      void  (IL   ???...  ???)
               [000022] ------------              \--*  JTRUE     void  
               [000020] ------------                 |  /--*  CNS_INT   int    0
               [000021] J------N----                 \--*  EQ        int   
               [000019] V-----------                    \--*  IND       int   
               [000018] ------------                       \--*  CNS_INT(h) int    0xF7AE7098 ptr

 poll block is BB02
               [000017] ------------              *  STMT      void  (IL   ???...  ???)
               [000016] --C-G-------              \--*  CALL help void   HELPER.CORINFO_HELP_POLL_GC

 bottom block is BB03
               [000007] ------------              *  STMT      void  (IL 0x000...0x004)
               [000003] --CXG-------              \--*  CALL      void   System.String.Concat
               [000001] ------------ arg0            +--*  LCL_VAR   ref    V00 arg0         
               [000002] ------------ arg1            +--*  LCL_VAR   ref    V01 arg1         
               [000008] ------------ arg2            +--*  CNS_INT   int    0
               [000010] ------------ arg3            +--*  CNS_INT   int    9
               [000012] ------------ arg4            +--*  CNS_INT   int    8
               [000014] ------------ arg5            \--*  CNS_INT   int    7


fgMorphTree BB01, stmt 1 (after)
               [000003] --CXG-------              *  CALL      void   System.String.Concat
               [000001] ------------ arg0         +--*  LCL_VAR   ref    V00 arg0         
               [000002] ------------ arg1         +--*  LCL_VAR   ref    V01 arg1         
               [000008] ------------ arg2         +--*  CNS_INT   int    0
               [000010] ------------ arg3         +--*  CNS_INT   int    9
               [000012] ------------ arg4         +--*  CNS_INT   int    8
               [000014] ------------ arg5         \--*  CNS_INT   int    7

Morphing BB02 of 'DynamicClass:MyMethod(ref,ref):ref'

fgMorphTree BB02, stmt 2 (before)
               [000016] --C-G-------              *  CALL help void   HELPER.CORINFO_HELP_POLL_GC
Initializing arg info for 16.CALL:

Morphing args for 16.CALL:


Morphing BB03 of 'DynamicClass:MyMethod(ref,ref):ref'

fgMorphTree BB03, stmt 3 (before)
               [000003] --CXG-------              *  CALL      void   System.String.Concat
               [000001] ------------ arg0         +--*  LCL_VAR   ref    V00 arg0         
               [000002] ------------ arg1         +--*  LCL_VAR   ref    V01 arg1         
               [000008] ------------ arg2         +--*  CNS_INT   int    0
               [000010] ------------ arg3         +--*  CNS_INT   int    9
               [000012] ------------ arg4         +--*  CNS_INT   int    8
               [000014] ------------ arg5         \--*  CNS_INT   int    7
Initializing arg info for 3.CALL:

Assert failure(PID 20524 [0x0000502c], Thread: 20524 [0x502c]): Assertion failed 'NYI: Morphing Vararg call not yet implemented on non Windows targets.' in 'DynamicClass:MyMethod(ref,ref):ref' (IL size 10)

    File: /media/kbaladurin/data/dotnet/forked/coreclr-sec/src/jit/morph.cpp Line: 2629
    Image: /media/kbaladurin/data/dotnet/clr_overlay_x86/corerun

Aborted (core dumped)
```

`fgMorphTailCall` marks `3.CALL` as varargs on x86 and amd64 and compilation fails due to NYI in `fgInitArgInfo` (on amd64 `fgMorphTailCall` isn't called as fast tail-call optimization is used). I think we can temporary disable tail-call optimization for Linux/x86:
```diff
diff --git a/src/jit/morph.cpp b/src/jit/morph.cpp
index 5f675dd76b..ddd5288ca6 100644
--- a/src/jit/morph.cpp
+++ b/src/jit/morph.cpp
@@ -8227,6 +8227,11 @@ GenTree* Compiler::fgMorphCall(GenTreeCall* call)
                 }
             }
         }
+#elif defined(_TARGET_UNIX_)
+        if (!canFastTailCall && szFailReason == nullptr)
+        {
+            szFailReason = "Morphing Vararg call that needed for tailcall optimization on non Windows targets is not yet implemented.";
+        }
 #endif // !_TARGET_X86_
 
         if (szFailReason != nullptr)
```

Thank you!

### PR DESCRIPTION
**Pull request by jashook**
Fixes #25020

### CODE REVIEW

**Comment by jashook** [comment]
This will standardize linux x86 and linux x64 to fail to explicitly tail call, because the jitee interface will return nullptr when asked for the tail stub.

ptal @BruceForstall @kbaladurin
/cc @dotnet/jit-contrib 

@kbaladurin can you verify the fix. I cannot build corefx for linux x86 to actually verify the test passes.

Note this adds a noway assert for _TARGET_UNIX_ fgMorphTailCall. This is to avoid the incorrect/unsupported vararg morph if we fail to correctly decide not to go the slow tailcall route.

**Comment by jashook** [comment]
Also @kbaladurin if you have updated information on how you are building and running linux x86 I would love to see it. Specifically, I had trouble building corefx and setting up the CORE_ROOT.

File: src/jit/morph.cpp [hunk 1]
```diff
@@ -7351,6 +7351,10 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee)
  */
 void Compiler::fgMorphTailCall(GenTreeCall* call, void* pfnCopyArgs)
 {
+#if defined(_TARGET_UNIX_)
+    noway_assert("Slow tail calls not supported on non-windows platforms.");
```
Comment (BruceForstall) [inline src/jit/morph.cpp]: This will never fail; you need a preceding `!`.

And please capitalize `Windows`.

**Comment by BruceForstall** [comment]
The approach seems fine, if it fixes the problem.

File: src/jit/morph.cpp (hunk 1 above)
Comment (jashook) [reply src/jit/morph.cpp]: Thanks changing

**Comment by kbaladurin** [comment]
@jashook thank you very much! I've tested this PR and it fixes the problem.

I build test coreroot using following way:
- build coreclr x86
```
# ./cross/build-rootfs.sh x86 bionic lldb4.0
$ ./build.sh cross x86
```
- build corefx x86 (native) and x64 (managed)
```
# ./cross/build-rootfs.sh x86 bionic lldb4.0
$ ./src/Native/build-native.sh cross x86
$ ./build.sh /p:BinPlaceNETCoreAppPackage=true
```
- copy coreclr and corefx binaries to coreroot:
```
$ cp <corefx>/artifacts/bin/native/Linux-x86-Debug/* <coreroot>
$ cp <corefx>/artifacts/bin/pkg/netcoreapp/runtime/* <coreroot>
$ cp <coreclr>/bin/Product/Linux.x86.Debug/* <coreroot>
```

**Comment by kbaladurin** [comment]
@viewizard FYI

**Comment by kbaladurin** [comment]
Is it ready to merge?

**Comment by jashook** [comment]
@kbaladurin this is ready to merge, I am waiting for master to switch to vnext. See https://github.com/dotnet/coreclr/pull/23323#issuecomment-506015898.

**Pull request merged**